File: Leetcode100/leetcode100_python/TrappingRainWater2.py
from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        result = 0
        max_left, max_right = 0, 0
        left, right = 0, len(height) - 1
        while left < right:
            if height[left] < height[right]:
                max_left = max(max_left, height[left])
                result += max_left - height[left]
                left += 1
            else:
                max_right = max(max_right, height[right])
                result += max_right - height[right]
                right -= 1
        return result

if __name__ == "__main__":
    height = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.trap(height))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/trapping_rain_water/TrappingRainWater2.java
package main.java.com.jiang.leetcode100.trapping_rain_water;

import java.util.Scanner;
import java.util.Stack;

class Solution2 {
    public int trap(int[] height) {
        int result = 0;
        int n = height.length;
        int[] max_left = new int[n];
        int[] max_right = new int[n];

        for(int i = 0; i < n; i++) {
            if(i == 0) max_left[i] = height[i];
            else max_left[i] = Math.max(max_left[i - 1], height[i]);
        }
        for(int i = n - 1; i >= 0; i--) {
            if(i == n - 1) max_right[i] = height[i];
            else max_right[i] = Math.max(max_right[i + 1], height[i]);
        }

        for(int i = 0; i < n; i++)
            result += Math.min(max_left[i], max_right[i]) - height[i];

        return result;
    }
}

public class TrappingRainWater2 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] height = new int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            height[i] = Integer.parseInt(nums_str[i]);
        }

        System.out.println(new Solution2().trap(height));
    }
}